[IMP] scss: add section, improve preexisting content (#3093)
odoo/documentation: content/developer/howtos/scss_tips.rst

Check the code before creating a new class. Chances are that there is already a class or a tag doing exactly what you're looking for.

On top of that, Odoo relies on `Bootstrap <https://getbootstrap.com/docs/5.1/getting-started/introduction/>`_ (BS), one of the most complete CSS framework available. The framework has been customized in order to match Odoo design (both community and enterprise versions), meaning that you can use any BS class, directly in Odoo and achieve a visual result consistent with our UI.

.. warning::
   The fact that a class achieve the desired visual result doesn't necessarily mean that's the right one for the job. Be aware of classes triggering JS behaviors, for example.

.. warning::
   Take care about classes semantic. To apply a **button class** to a **title** is not just semantically wrong, it may also lead to migration issues and visual inconsistencies.

The following sections describe tips to strip-down SCSS lines **when custom-code is the only way to go**.

Browsers defaults
=================

By default, each browser renders content using an *user agent stylesheet*.

> To overcome inconsistencies, some of these rules are overridden by `Bootstrap Reboot <https://getbootstrap.com/docs/5.1/content/reboot/>`_.

Reviewer: ```suggestion
To overcome inconsistencies between browsers, some of these rules are overridden by `Bootstrap Reboot <https://getbootstrap.com/docs/5.1/content/reboot/>`_.
```

Author: Isn't that clear enough thanks to the prev. sentence?

Reviewer: I always think it's best to write as though the reader was stupid 😅
But it's not necessarily clear because the sentence can be understood in different ways.

   .. code-block:: css

      div.element {
         display: block;
         /* not needed 99% of the time */
      }

.. example::

   For the same reason, sometimes you may opt to switch html tag rather than adding a new CSS rule.

   .. code-block:: css

      span.element {
         display: block;
         /* replace <span> with <div> instead
            to get 'display: block' by default */
      }

Follows a non-comprehensive list of default rules:

.. list-table::
   :header-rows: 1

   * - Tag / Attribute
     - Defaults
   * - `<div/>`, `<section/>`,
       `<header/>`, `<footer/>` ...
     - `display: block`
   * - `<span/>`, `<a/>`
       `<em/>`, `<b/>` ...
     - `display: inline`
   * - `<button/>`, `<label/>`
       `<output/>`...
     - `display: inline-block`
   * - `<img/>`, `<svg/>`
     - `vertical-align: middle`
   * - `<summary/>`, `[role="button"]`
     - `cursor: pointer;`
   * - `<q/>`
     - `:before {content: open-quote}`
       `:after  {content: close-quote}`
   * - ...
     - ...

.. seealso::
   Bootstrap Reboot `on github <https://github.com/twbs/bootstrap/blob/1a6fdfae6b/scss/_reboot.scss>`_

> Use tags

Reviewer: ```suggestion
Using tags
```

   Following the last example, if Odoo titles' design changes, these changes will be applied in the `o_module_custom_title` element too since it's using a the `<h5>` tag.

Utility classes
===============

Our framework defines a multitude of utility classes designed to cover almost all layout/design/interaction needs.
The simple fact that a class exists justify its use over custom CSS whenever is possible.

Take the example of `position-relative`. What does is quite self-explicative, but *why* should be used maybe a little less.

.. code-block:: css

   position-relative {
      position: relative !important;
   }

Since an utility-class is defined, any CSS line with declaration `position: relative` is **potentially** redundant.

Odoo relies on the default `Bootstrap utility-classes <https://getbootstrap.com/docs/5.1/utilities/background/>`_ stack and define its own ones using `Bootstrap API <https://getbootstrap.com/docs/5.1/utilities/api/>`_.

.. seealso::
   | Bootstrap `utility classes <https://getbootstrap.com/docs/5.1/utilities/api/>`_
   | Odoo `custom utilities on github <https://github.com/odoo/odoo/blob/16.0/addons/web/static/src/scss/utilities_custom.scss>`_

> Handle utility-classes verbosity

Reviewer: ```suggestion
Handling utility-classes' verbosity
```

--------------------------------

> The downside of utility-classes is the potential lack of readability, eg:

Reviewer: ```suggestion
The downside of utility-classes is the potential lack of readability, e.g.:
```

.. code-block:: html

   <myComponent t-attf-class="d-flex border px-lg-2 card
   {{props.readonly ? 'o_myComponent_disabled' : ''}}
   card d-lg-block position-absolute {{props.active ?
   'o_myComponent_active' : ''}}  myComponent px-3"/>

To overcome the issue you may combine different approaches:

> - in qweb attributes, use classes to be toggled *on-the-fly* only

Reviewer: ```suggestion
- in QWeb attributes, only use classes to be toggled *on-the-fly*;
```

> - use new lines for each attribute

Reviewer: ```suggestion
- use new lines for each attribute;
```

> - order classes using the convention `[odoo component] [bootstrap component] [css declaration order]`

Reviewer: ```suggestion
- order classes using the convention `[odoo component] [bootstrap component] [css declaration order]`.
```
